hazelcast/src/main/java/com/hazelcast/client/impl/connection/tcp/TcpClientConnection.java

```diff
@@ -106,45 +108,30 @@ public TcpClientConnection(HazelcastClientInstanceImpl client, int connectionId)
         this.connectionId = connectionId;
         this.channel = null;
         this.attributeMap = null;
+        this.channelInitializer = null;
         this.logger = client.getLoggingService().getLogger(TcpClientConnection.class);
     }
 
-    public ChannelInitializer getChannelInitializer() {
-        return channelInitializer;
-    }
-
-    public void setChannelInitializer(ChannelInitializer channelInitializer) {
-        this.channelInitializer = channelInitializer;
-    }
-
-    public void setTpcChannels(Channel[] tpcChannels) {
-        this.tpcChannels = checkNotNull(tpcChannels);
-    }
-
     @Override
     public ConcurrentMap attributeMap() {
         return attributeMap;
     }
 
     @Override
     public boolean write(OutboundFrame frame) {
+        Channel[] altoChannels = this.altoChannels;
+        if (altoChannels == null) {
+            return channel.write(frame);
+        }
+
         ClientMessage clientMessage = (ClientMessage) frame;
         int partitionId = clientMessage.getPartitionId();
-        if (tpcChannels == null || partitionId < 0) {
-            if (channel.write(frame)) {
-                return true;
-            }
-        } else {
-            int tpcChannelIndex = HashUtil.hashToIndex(partitionId, tpcChannels.length);
-            if (tpcChannels[tpcChannelIndex].write(frame)) {
-                return true;
-            }
+        if (partitionId < 0) {
+            return channel.write(frame);
         }
 
-        if (logger.isFinestEnabled()) {
-            logger.finest("Connection is closed, dropping frame -> " + frame);
-        }
-        return false;
+        int channelIndex = partitionId % altoChannels.length;
```

Comment:
Can partitionId be negative?

Reply:
For partition-specific invocations, no. Also, note that there is a check above with the partition id. If it is negative, we will use the legacy channel